fuel-vm/src/storage/predicate.rs

```
@@ -63,246 +92,281 @@ impl From<StorageUnavailable> for RuntimeError<StorageUnavailable> {
-    ) -> Result<Option<usize>, StorageUnavailable> {
-        Err(StorageUnavailable)
+    ) -> Result<Option<usize>, Self::Error> {
+        Err(Self::Error::UnsupportedStorageOperation)
     }
 
     fn read_alloc(
         &self,
         _key: &<ContractsRawCode as Mappable>::Key,
-    ) -> Result<Option<Vec<u8>>, StorageUnavailable> {
-        Err(StorageUnavailable)
+    ) -> Result<Option<Vec<u8>>, Self::Error> {
+        Err(Self::Error::UnsupportedStorageOperation)
     }
 }
 
-impl StorageWrite<ContractsRawCode> for PredicateStorage {
+impl<D> StorageWrite<ContractsRawCode> for PredicateStorage<D>
+where
+    D: StorageRead<BlobData>,
+{
     fn write_bytes(
         &mut self,
         _key: &<ContractsRawCode as Mappable>::Key,
         _buf: &[u8],
     ) -> Result<usize, Self::Error> {
-        Err(StorageUnavailable)
+        Err(Self::Error::UnsupportedStorageOperation)
     }
 
     fn replace_bytes(
         &mut self,
         _key: &<ContractsRawCode as Mappable>::Key,
         _buf: &[u8],
     ) -> Result<(usize, Option<Vec<u8>>), Self::Error> {
-        Err(StorageUnavailable)
+        Err(Self::Error::UnsupportedStorageOperation)
     }
 
     fn take_bytes(
         &mut self,
         _key: &<ContractsRawCode as Mappable>::Key,
     ) -> Result<Option<Vec<u8>>, Self::Error> {
-        Err(StorageUnavailable)
+        Err(Self::Error::UnsupportedStorageOperation)
     }
 }
 
-impl StorageSize<ContractsState> for PredicateStorage {
+impl<D> StorageSize<ContractsState> for PredicateStorage<D>
+where
+    D: StorageRead<BlobData>,
+{
     fn size_of_value(
         &self,
         _key: &<ContractsState as Mappable>::Key,
-    ) -> Result<Option<usize>, StorageUnavailable> {
-        Err(StorageUnavailable)
+    ) -> Result<Option<usize>, Self::Error> {
+        Err(Self::Error::UnsupportedStorageOperation)
     }
 }
 
-impl StorageRead<ContractsState> for PredicateStorage {
+impl<D> StorageRead<ContractsState> for PredicateStorage<D>
```

question for my own understanding: in what cases can these operations be called?

---

We allow only blobs in the predicate. Any other access to the storage should return an error